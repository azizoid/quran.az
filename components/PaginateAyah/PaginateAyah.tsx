import Link from 'next/link'

export type PaginateAyahProps = {
  soorah: number
  ayah: number
  prev?: number
  next?: number
  translator?: number
}

export const PaginateAyah = ({
  soorah,
  ayah,
  prev,
  next,
  translator,
}: PaginateAyahProps): JSX.Element => (
  <div className="pagination">
    {prev && (
      <Link href={`/${soorah}/${prev}?t=${translator}`}>
        <a className="pagination-item">{prev}</a>
      </Link>
    )}
    <span className="pagination-disabled">{ayah}</span>
    {next && (
      <Link href={`/${soorah}/${next}?t=${translator}`}>
        <a className="pagination-item">{next}</a>
      </Link>
    )}
  </div>
)
