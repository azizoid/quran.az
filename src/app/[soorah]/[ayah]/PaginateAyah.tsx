import { ChevronLeftIcon, ChevronRightIcon } from 'lucide-react'

import { soorahList } from '@/assets/soorah-list-object'
import {
  Pagination as PaginationShadcn,
  PaginationContent,
  PaginationEllipsis,
  PaginationItem,
  PaginationLink,
} from '@/components/ui/pagination'
import { buildUrl } from '@/helpers/buildUrl'
import { calculatePageBounds } from '@/utility/calculatePageBounds/calculatePageBounds'

export type PaginateAyahProps = {
  soorah: number
  activePage: number
  pageRangeDisplayed?: number
}

export const PaginateAyah = ({ soorah, activePage, pageRangeDisplayed = 5 }: PaginateAyahProps) => {
  const totalPages = soorahList[soorah - 1].ayahCount

  const { startPage, endPage } = calculatePageBounds(activePage, totalPages, pageRangeDisplayed)
  const pageNumbers = Array.from({ length: endPage - startPage + 1 }, (_, i) => startPage + i)

  return (
    <PaginationShadcn>
      <PaginationContent>
        {activePage > 1 && !pageNumbers.includes(1) && (
          <PaginationItem>
            <PaginationLink href={buildUrl(soorah, activePage - 1)}>
              <ChevronLeftIcon className="h-4 w-4" />
            </PaginationLink>
          </PaginationItem>
        )}

        {pageNumbers.map((page) => {
          const isCurrPage = page === activePage

          return (
            <PaginationItem key={page}>
              <PaginationLink href={buildUrl(soorah, page)} isActive={isCurrPage}>
                {page}
              </PaginationLink>
            </PaginationItem>
          )
        })}

        {!pageNumbers.includes(activePage) && (
          <PaginationItem>
            <PaginationEllipsis />
          </PaginationItem>
        )}

        {activePage !== totalPages && !pageNumbers.includes(totalPages) && (
          <PaginationItem>
            <PaginationLink href={buildUrl(soorah, activePage + 1)}>
              <ChevronRightIcon className="h-4 w-4" />
            </PaginationLink>
          </PaginationItem>
        )}
      </PaginationContent>
    </PaginationShadcn>
  )
}
