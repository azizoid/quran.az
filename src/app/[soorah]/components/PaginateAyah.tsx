import { ChevronLeftIcon, ChevronRightIcon } from 'lucide-react'

import { soorahList } from '@/assets/soorah-list-object'
import {
  Pagination as PaginationShadcn,
  PaginationContent,
  PaginationEllipsis,
  PaginationItem,
  PaginationLink,
} from '@/components/ui/pagination'
import { buildUrl } from '@/helpers/buildUrl'
import { calculatePageBounds } from '@/utility/calculatePageBounds/calculatePageBounds'

export type PaginateAyahProps = {
  soorah: number
  activePage: number
  translator: number
  pageRangeDisplayed?: number
}

export const PaginateAyah = ({
  soorah,
  activePage,
  translator,
  pageRangeDisplayed = 5,
}: PaginateAyahProps) => {
  const totalPages = soorahList[soorah - 1].ayahCount

  const { startPage, endPage } = calculatePageBounds(activePage, totalPages, pageRangeDisplayed)
  const pageNumbers = Array.from({ length: endPage - startPage + 1 }, (_, i) => startPage + i)

  return (
    <PaginationShadcn>
      <PaginationContent>
        {activePage > 1 && !pageNumbers.includes(1) && (
          <PaginationItem>
            <PaginationLink href={buildUrl(soorah, activePage - 1, translator)}>
              <ChevronLeftIcon className="h-4 w-4" />
            </PaginationLink>
          </PaginationItem>
        )}

        {pageNumbers.map((page) => {
          const isCurrPage = page === activePage

          return (
            <PaginationItem key={page}>
              <PaginationLink href={buildUrl(soorah, page, translator)} isActive={isCurrPage}>
                {page}
              </PaginationLink>
            </PaginationItem>
          )
        })}

        {!pageNumbers.includes(activePage) && (
          <PaginationItem>
            <PaginationEllipsis />
          </PaginationItem>
        )}

        {activePage !== totalPages && !pageNumbers.includes(totalPages) && (
          <PaginationItem>
            <PaginationLink href={buildUrl(soorah, activePage + 1, translator)}>
              <ChevronRightIcon className="h-4 w-4" />
            </PaginationLink>
          </PaginationItem>
        )}
      </PaginationContent>
    </PaginationShadcn>
  )
}
