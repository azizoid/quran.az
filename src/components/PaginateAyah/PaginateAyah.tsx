import { soorahList } from '@/assets/soorah-list-object'

import { PaginationLink } from '../PaginationLink/PaginationLink'

export type PaginateAyahProps = {
  soorah: number
  ayah: number
  translator?: number
}

export const PaginateAyah = ({ soorah, ayah, translator }: PaginateAyahProps) => {
  const soorahIndex = soorah - 1

  const prevAyah = ayah === 1 ? null : ayah - 1
  const nextAyah = ayah === soorahList[soorahIndex]?.ayahCount ? null : ayah + 1

  const prevSoorah = !prevAyah && soorah > 1 ? soorahList[soorahIndex - 1]?.fullTitle : null
  const nextSoorah = !nextAyah && soorah < 114 ? soorahList[soorahIndex + 1]?.fullTitle : null

  return (
    <div className="pagination">
      {prevSoorah && (
        <PaginationLink href={`/${soorah - 1}?t=${translator}`}>{`${
          soorah - 1
        }. ${prevSoorah} ←`}</PaginationLink>
      )}
      {prevAyah && (
        <PaginationLink href={`/${soorah}/${prevAyah}?t=${translator}`}>{prevAyah}</PaginationLink>
      )}

      <span className="pagination-disabled">{ayah}</span>

      {nextAyah && (
        <PaginationLink href={`/${soorah}/${nextAyah}?t=${translator}`}>{nextAyah}</PaginationLink>
      )}
      {nextSoorah && (
        <PaginationLink href={`/${soorah + 1}?t=${translator}`}>{`${
          soorah + 1
        }. ${nextSoorah} →`}</PaginationLink>
      )}
    </div>
  )
}
