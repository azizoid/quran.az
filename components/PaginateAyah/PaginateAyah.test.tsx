import { render } from '@testing-library/react'
import { PaginateAyah } from './PaginateAyah'

test('PaginateAyah Snapshot', async () => {
  const { container, getByText } = render(<PaginateAyah soorah={1} ayah={5} prev={4} next={6} />)

  expect(getByText('4')).toBeInTheDocument()
  expect(getByText('5')).toBeInTheDocument()
  expect(getByText('6')).toBeInTheDocument()

  expect(container).toMatchSnapshot()
})

test('PaginateAyah middle of ayah', async () => {
  const { queryByText } = render(<PaginateAyah soorah={1} ayah={7} prev={6} />)

  expect(queryByText('8')).not.toBeInTheDocument()
})
