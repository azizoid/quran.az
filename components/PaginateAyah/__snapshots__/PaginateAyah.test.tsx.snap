// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginateAyah Snapshot 1`] = `
<div>
  <div
    class="pagination"
  >
    <a
      class="pagination-item"
      href="/1/4?t=undefined"
    >
      4
    </a>
    <span
      class="pagination-disabled"
    >
      5
    </span>
    <a
      class="pagination-item"
      href="/1/6?t=undefined"
    >
      6
    </a>
  </div>
</div>
`;
